### ISSUE TITLE
Question: more functionality on specs Camera

### ISSUE THREAD

**Issue by derekdreery** [amethyst/amethyst#176]
I want to quickly implement a free-flow camera, and to do this I need to be able to rotate the view. The specs camera makes this difficult (you'd need to move the lookat value in a circle around the eye).

I'm happy to implement some methods on the Camera (something like below) if the project owners think this is valuable?

```rust
impl Camera {
    /// Rotates the camera relative to the current view direction
    pub fn view_rotate(&mut self, pitch: f32, yaw: f32, roll: f32) { ... }

    /// Rotates the camera relative to the world coordinate system
    pub fn world_rotate(&mut self, x: f32, y: f32, z: f32) { ... }
}
```

**Comment by OvermindDL1** [comment]
If you want to rotate relative then taking a quaternion instead of pitch/yaw/roll or x/y/z would remove ambiguities in the transformation (especially if you want to lerp it over time).

**Comment by derekdreery** [comment]
Ok I've read some literature and see why quaternions are useful. Maybe the API could be:

```rust
impl Camera {
    /// Rotate camera using given quaternion
    pub fn rotate_quaternion(&mut self, q: cgmath::Quaternion<f32>) { ... }

    /// Rotate camera about given axis by given amount (clockwise)
    pub fn rotate(&mut self, angle: f32, axis: cgmath::Vector3<f32>) { ... }
}

**Comment by derekdreery** [comment]
Or could take a `[f32; 4]` instead of a quaternion for compatibility.

**Comment by OvermindDL1** [comment]
> Or could take a `[f32; 4]` instead of a quaternion for compatibility.

That would not have guaranteed alignment though, a nice thing about quaternions as well is the SSE'abilities of them.  Plus it is documentative.  :-)

I like the two you shown though, a `rotate_quaternion` method is the one that shoul dbe used, but your `rotate` method taking an angle and an axis (really need vectors for documentative purposes as well I'd say...) is also not bad, but it does calculate slower than a quaternion (a quaternion is just a 4d unit vector, so you can use traditional vector optimizations on them, including SSE commands).

However, in my opinion a `rotate` function should not take alternatives, it should take 'one source of truth', and instead the quaternion should have a helper on it that takes an angle and axis to build a quaternion from that.

**Comment by ebkalderon** [comment]
@derekdreery I think this might be a better question for the `specs` developers to handle, since we are essentially using their `Camera` resource verbatim (discounting the completely separate `Camera` in the renderer crate which serves an entirely different purpose).

However, I think it would be doable to expose a custom trait and implement it against the `Camera` type which adds the additional behavior you want. What does everyone think of this from a technical perspective? User experience as a developer? Are there shortcomings or possible issues with this approach? Should we just delegate these suggestions to the `specs` project and let them deal with it?

**Comment by Aceeri** [comment]
I'm pretty sure we define `Camera` though? `specs` just gives us the ability to add resources.

**Comment by ebkalderon** [comment]
@Aceeri Are you sure? I'll quickly check.

**Comment by Aceeri** [comment]
I'm more of the opinion on leaving default methods in and allowing users to use their own math libraries (cgmath, nalgebra, etc.) One thing we should discuss though is what the structure of the quaternion and such should look like.

**Comment by msiglreith** [comment]
I would prefer hiding the `Camera` struct a bit more and encourage the users to use a combination of `Transform` and `CameraComponent` (doesn't exist so far!). `CameraComponent` would then store information for perspective projection and other related lens settings (e.g shutter speed, ..) -> `proj` matrix.
The `view` matrix will be automatically filled using the inverse of the `Transform` matrix.

`look_at` and `rotate` could then be implemented for `Transform`.

**Comment by derekdreery** [comment]
I'll just give some background so it's clear why I suggested this.

I am a developer that knows rust, but not amethyst. I want to be able to explore the game world easily as specs runs a simulation. To do this I'm implementing WASD as forward/back/strafe, and mouse right click + drag to change camera angle. This is for example how the asset viewer in Unreal works. To do this I've needed to add a mouse state handler similar to the keyboard handler, and now I need to be able to rotate the camera relative to the view coordinate system, by which I mean, when the mouse is moved

 - get the mouse move `(horiz, vert)`
 - get the look direction (look_at - eye)
 - make a 3d vector `move = [horiz, vert, 0]`, no rotation along view axis
 - get the amount of rotation (`|move| = sqrt(dot(move, move))`) maybe multiplied by some constant
 - get the axis of rotation (`cross_product(move/|move|, [0, 0, 1])`) quick because can treat as 2d cross
 - get the transformation matrix T for view space (need to work this out)
 - perform transformation `TqT^-1` where q is the rotation quaternion to get new view matrix (ignoring world transform is inverse camera transform) (not sure how this works with eye, look_at, and up)

Having thought about it, I guess the camera struct is just getting in the way of what I want to do here, I'd rather just work with the view matrix. I think not having to understand the view matrix in order to use amethyst is a good thing though, I could just create the camera and add it to the system.

Maybe the system could be designed as @msiglreith says, but then there is some sort of `SimpleCameraOrientation` that implements `Transform` and works like `Camera`, so new users can just use this (plus docs so new users can get something on screen easily).

This comment is a bit of a stream of conciousness, I hope it adds something to the discussion.

**Comment by msiglreith** [comment]
Follow-up for my initial proposal:
```Rust
pub struct FpsCameraController;
pub struct FpsCameraComponent; // marker component only

impl System<()> for FpsCameraController {
   fn run(&mut self, ctx: &()) {
       // Gather input from `InputHandler`
       // Translate input state into camera actions (rotate, translate, etc.)
       // Update `Transform` or `CameraComponent` of the all cameras with the marker component
   }
}
...

impl System<()> for Renderer {
    fn run(...) {
        // Note: doesn't need to know about different camera models!
        for (camera, transform) in (&cameras, &transforms).iter() {
           let render_camera = Camera { proj: camera.get_projection(), view: inverse(transform.get_world_transform()) };
           scene.add_camera(render_camera); // or only active camera..
        }
    }
}
```
`FpsCameraController` would be an optional system, mainly for users to get started as mentioned by @derekdreery. For different games (e.g 2d topdown), other camera models can be implemented which differ in input handling. Therefore users only need to deal with transforming their camera based on input actions and don't have to deal with view or projection matrices directly.
Re-using the `Transform` component also allows us to embed it easier into future animation systems for example.

**Comment by derekdreery** [comment]
That looks sensible to me. You could also have StaticCameraController for the current Camera functionality, and the user could implement their own if they prefer.

This also means you could have cameras in 3rd party modules that downstream can use off the shelf, rather than having to implement every possible camera system in the `amethyst` lib directly.

Would you have some sort of configuration for FpsCameraController (e.g. not hard-coding which keys do what, do you need to press a mouse button to rotate the camera or not?, which button?, do you move towards the camera (flying) or stay on the ground like an FPS game?)

**Comment by derekdreery** [comment]
Maybe the flying camera is sufficiently different from the FPS camera that it should be separate.

**Comment by msiglreith** [comment]
A free view camera might be indeed the best controller to implement atm as it offers large control freedom (3 DoF for translation and 2/3 DoF for orientation). I only used a fps camera for the example.

For key mapping https://github.com/amethyst/amethyst/issues/177 would be nice to have, but we could also store the key mapping directly in the `Controller` struct for the moment. e.g:
```Rust
pub struct Controller {
    pub forward: VirtualKeyCode,
    pub left: VirtualKeyCode,
    ...
}
```

**Comment by derekdreery** [comment]
I'd like to help implementing this as it's blocking what I want to do in usercode.

My main question/comment is that camera movement in general depends on much more of the global state than user input (e.g. clipping, getting shot and falling over etc.), so in general this will be implemented a lot in user code. Therefore it needs to be easy to understand by users. That's probably just a documentation thing, doesn't impact coding, but I think to implement your own camera you will have to understand 3d math.

Should there be an example dedicated to cameras?

**Comment by derekdreery** [comment]
Another question: should the camera transformation be called `View` to match MVP, or `CameraView`, or `ViewTransformation`, `Transformation`, or something else?

**Comment by msiglreith** [comment]
> I'd like to help implementing this as it's blocking what I want to do in usercode.

Great! Looking forward to the implementation.

> Should there be an example dedicated to cameras?

I would prefer implementing a free view camera in one of the existing examples.

> Another question: should the camera transformation be called View to match MVP, or CameraView, or ViewTransform, Transform, or something else?

Amethyst is highly designed around the concept of ECS (entity-component-system). So a camera will only be a entity, which is defined to have a `Transform` and `CameraComponent` component, which are stored in their own storages.

> My main question/comment is that camera movement in general depends on much more of the global state than user input (e.g. clipping, getting shot and falling over etc.), so in general this will be implemented a lot in user code. Therefore it needs to be easy to understand by users. That's probably just a documentation thing, doesn't impact coding, but I think to implement your own camera you will have to understand 3d math.

True! We maybe can aid developers by providing contraints and collision detection function, but that's out of scope for this issue imo.

**Comment by derekdreery** [comment]
OK thanks for the info about the ecs model. 

What I meant by "exampe for cameras" really was "example of how to implement camera using Transform/CameraComponent", which could just be in the API docs.

**Comment by minecrawler** [comment]
Is anyone working on the camera at the moment? Since I am diving into Amethyst 3D, a sane camera API is essential. At the moment, I can easily simulate left/right/forward/backward by moving the camera and changing the lookAt position at the same time. However, adding rotation is a lot harder, and personally, I think that a game engine should provide easy to use procedures (like `camera.rotateByDegree(x,y,z)`, maybe making use of a "rotate-able" component).

I tried to make sense of the camera implementation in order to add some code (implement quaternions, add helper procedures), however I found code in at least two different locations and I don't know how they are connected and in what way I can change the render code in order to manipulate the necessary vectors and matrices.

Could anyone give me a bit of advice on how to start changing the camera code?

**Comment by derekdreery** [comment]
The reason I stopped looking at this was because I didn't really understand the renderer, and that was required to make changes to the MVP matrices.

The renderer is undergoing a major rewrite at the moment, so it's probably best to hold back until that lands  ( #233 ).

**Comment by Binero** [comment]
I think the best way to go is to just have the Camera structure be as flexible as possible, a view matrix or something very close to it. Specific cameras like an FPS camera could then be added in crates. (e.g. they could introduce a system to setup the view matrix)

**Comment by minecrawler** [comment]
@Binero just a bit further up, we discussed that a game-agnostic Quaternion-based camera, which exposes convenience-functionalities like `rotate_by_degree()`, might be a good idea. Especially in the early stages, it is very important to get the basics working and not think a lot about specifics, so that we have rapid iterations and an MVP for 2D and 3D based game development. Is that satisfactory for you or do you suggest any changes?

**Comment by Binero** [comment]
@minecrawler I did read about those convenience methods, but I dismissed them. Different types of cameras work best with different types of interfaces to them. I think it makes more sense to have different types of cameras, which uses things like yaw, pitch, rotation, eye, or a third person camera which just uses the eye, target and up. I just don't think just convenience methods will cut it, and as amethyst already uses a structure containing a view and a projection matrix internally, I see no harm in exposing it and allowing direct manipulation.

**Comment by Binero** [comment]
To elaborate on my above comment, that is not to say the convenience methods should not be available. I just think it's important to also allow low level manipulation for those who are comfortable with it.

**Comment by derekdreery** [comment]
@Binero I think the plan is to implement the camera as view/projection matrices (or some other representation e.g. quaternions), and then provide some default cameras, implementing a trait with methods like `projection_matrix(&self) -> &SomeType` or `view_quaternion(&self) -> &SomeQuaternionType`. This way beginners can just plug in a camera, whereas users that want control implement the trait for their own struct (maybe there is an impl for `Matrix4`/`Quaternion` so you can just use these.

**Comment by derekdreery** [comment]
There are 2 equally valid use-cases here
 1. Beginner wanting to focus on making models and be able to move around them easilly, or someone wanting to focus on shaders, not game logic/camera position.
 2. Game writer who needs control of view/projection matrices.

**Comment by OvermindDL1** [comment]
As long as the camera transformation matrix is exposed then we can build whatever helpers we want on top of it.

**Comment by Binero** [comment]
@derekdreery I don't think reimplementing quaternions is a good idea. People who do not want to deal with the raw matrices will surely prefer a camera-interface? And similarily, someone who doesn't want a camera interface would want to be able to change the view matrix using their own crate, like `cgmath` or `nalgebra`. I think implementing a new quaternion struct just for the camera is reinventing the wheel, with no particular use cases. (but perhaps annoyance by the advanced users who want to use their likely more-flexible nalgebra or cgmath interface)

**Comment by minecrawler** [comment]
@Binero I think we would rather use cgmath, since we already use that for other parts, too, though

**Comment by derekdreery** [comment]
I was being very vague in my above comment, I didn't mean to create new types when existing types in cgmath or libstd work fine. :)

**Comment by kvark** [comment]
I encourage you to expose matrices, quaternions, and whatever else in [Mint](https://github.com/kvark/mint) types, see #236

**Comment by Binero** [comment]
Although I am not sure what use case exposing a quaternion/translation has, as long as it allows math crate-agnostic usage, I guess it could be a compromise for exposing the view matrix directly.

**Comment by derekdreery** [comment]
I think just expose whatever format you are going to send to the graphics card.

**Comment by msiglreith** [comment]
Something which should be done beforehand is to discuss the camera representation imo.
There currently seems to be 2 representation on how to the concept of a camera should be integrated into the engine:
* Resource object: One/multiple cameras are added as resources to the world with some properties. Users can call functions on these cameras (e.g move/rotate/...)
* Component based: Cameras are entities with components for camera properties

I'm leaning towards the component based approach as the concept of a camera is more flexible here. With 'concept of camera' I mean that systems have different requirements on how a camera should be represented in terms of properties: renderer wants view matrix, proj matrix and maybe some lens parameters. Game designer might have some more game specific parameters for animation.
A challenge here is to find a sensible split between properties and how to reuse components and minimize costs of transformation between representation (e.g 'player' camera into camera for scene rendering).

That sounds pretty abstract so in practice I'm asking for how for example the view matrix is stored:
Renderer wants a view matrix, which might be not that suitable representation for the game logic. Also, do we actually store the view matrix or reconstruct it from `Transform`, having a higher CPU cost but more flexbile.

Going back to the functionality aspect: Traits don't work that well with a component based approach. If we want to use these as interface resource objects will be more suitable imo.

**Comment by OvermindDL1** [comment]
> Component based: Cameras are entities with components for camera properties

Absolutely this one, this is what I do in my C++ engine and it just has an associated framebuffer for each camera in the components (by a framebuffer name mapping, which if not implemented elsewhere just becomes a no-op camera, good for turning on and off).  I can attach a camera component to anything with a transformation matrix on the rendering viewport to offset from the base position as well.

**Comment by gaudecker** [comment]
I looked at my personal requirements for the camera, and tried to apply them for its representation and implementation in Amethyst.

So far my requirements seem to support representing the cameras as their own entities:

- Attaching the camera to a path, and interpolating its position and orientation
  along the path.
  
  For cinematic shots and cutscenes. The path interpolator can just modify the
  standard transform component of the camera.
  
- Attaching the camera to other entities or a point in space.

  An obvious requirement and easily achieved by having the camera be its own
  entity.
  
- Multiple cameras and render to texture.

  Common effect in games for rendering a real-time TV broadcast or security
  camera feed for example. Split-screen multiplayer.

- Interpolation between projections.

  Interpolating between orthographic and perspective projection can make a cool
  special effect. With the projection matrix being in a component, it can be
  easily modified.

A camera being just another entity makes it easy to script cutscenes and sequences.

**Comment by Binero** [comment]
How would one create a camera entity? Would a camera component not be more suitable? A camera would then be described by any entity that has both a `Transform` and a `Camera`-flag component set. This also means we can re-use existing code from `Transform`, and do things like attaching cameras to objects in the scene. 

The obvious downside is that it might be confusing that a `Camera` also needs a `Transform` set.

**Comment by torkleyy** [comment]
I'm pretty sure @gaudecker was talking about just this.

**Comment by Binero** [comment]
Would it be acceptable to for now:

- Move `Camera` out of resources into its own component.
- Make `Camera` only a wrapper around `Projection`, and thus requiring a `Transform` to work properly. 
- Render nothing if there is no entity that has a `Camera` and a `Transform`.
- Panic if there are multiple entities that have a `Camera` and a `Transform`.

**Comment by omni-viral** [comment]
@Binero 
There is a cases when you need multiple cameras. 
I think `Camera` should be a component. But consist of `Projection`, render target and possibly enabled flag.

**Comment by Binero** [comment]
@omni-viral Are we anywhere near close to allowing multiple `Camera`s though? I would like to see an easier to use `Camera` then the current one as fast as possible. Furthermore, I think allowing multiple `Camera`s is certainly a must later on, but it might be out of the scope of this issue, which simply asked for a nicer interface.

**Comment by omni-viral** [comment]
@Binero I've just pointed that multi-camera support is necessary. Of course implementation for real multi-camera can be added later. But it should be considered during requested rewrite of interface.
So `Camera` may be provided with methods to bind to render target while others than binding to default one are left unimplemented.

**Comment by OvermindDL1** [comment]
Such a Camera component really should have a data attribute on it that points to a framebuffer of some sort (could just be a name or an index or something, I use an Atom64 in my engine for it, which is a string that is a unique integer at compile-time).  That allows multiple cameras to render based on the framebuffer that needs to be rendered at the current time (one of which just goes straight to the screen of course).  You can even have multiple cameras render to the same framebuffer if you give them a priority (and of course that 'later' ones don't render some parts of it to let earlier parts 'show through'), that is quite useful of an ability as well, one example of which is putting a camera in a mini-scene that renders first then the user camera renders (after clearing the depth buffer only, or if the first camera does not render depth at all if the scene allows for it), this would be used for example at rendering a fake 'far far' away (clouds, city, whatever) that moves only minorly based on how the user moves for translation (but rotates normally), but the main camera runs and jumps and does whatever they want.

In my engine I've done things where I have 6 cameras that render chunks of distance, say, 6+ (with each chunk being 32x32x32 units in size, say meters) only once per time that the user 'changes' the chunk they are in (by just moving), and it renders once to a framebuffer, and that framebuffer is blitted to the screen framebuffer at the start of each frame render, that way I can render *far* away things like mountains and so forth (I don't render things move a lot at that distance unless they are 'really' large, in which cause they themselves cause a far-distance render update of just the camera's that include them) but only once in a while, where the main user camera moves in their own area normally.

Even those are only the start of tricks I've done with cameras.  ^.^;

**Comment by omni-viral** [comment]
@OvermindDL1 Sophisticated tricks. But I believe there shouldn't be cameras involved.
In my opinion there should be one camera for each viewport. Viewports can be linked to same render target or portion of it.

**Comment by OvermindDL1** [comment]
> viewport

What defines a `viewport` though, usually in my system that is just a framebuffer (or perhaps even a portion of it).  Or are you wanting to split the camera/view distinction into two things (I've always kept it as just one 'thing').

**Comment by omni-viral** [comment]
In my view of engine architecture `Camera` is just transformation matrix. `Viewport` is pointer to render target, rectangle to render into, depth to prioritize with other viewports, possibility camera id to transform coordinates.

**Comment by OvermindDL1** [comment]
> In my view of engine architecture Camera is just transformation matrix. Viewport is pointer to render target, rectangle to render into, depth to prioritize with other viewports, possibility camera id to transform coordinates.

Ah, quite different from mine.  In mine I generally had the Camera be a pointer to a render target, and the transformation matrix was just whatever transformation matrix component was on the same node that the camera component happened to be on (without a transformation component it would not render 'world' items but could do custom rendering if other components were on, used for screenspace menu's and such).

**Comment by omni-viral** [comment]
@OvermindDL1 transformation component usually holds transform from local to world space. Camera needs it inverted + projection.

**Comment by Xaeroxe** [comment]
Fixed in #447
